File: application/src/test/java/com/smtm/application/MvcStepdefs.java
package com.smtm.application;

import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.content;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.header;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
import java.util.Map;
import org.json.JSONObject;
import org.springframework.http.HttpHeaders;
import org.springframework.test.web.servlet.MockMvc;
import org.springframework.test.web.servlet.ResultActions;
import io.cucumber.java.en.Given;
import io.cucumber.java.en.Then;
import io.cucumber.java.en.When;

public class MvcStepdefs {

    private final MockMvc mvc;
    private final HttpHeaders headers;

    private ResultActions actions;

    public MvcStepdefs(MockMvc mvc) {
        this.mvc = mvc;
        this.headers = new HttpHeaders();
    }

    @Given("request headers are")
    public void requestHeadersAre(Map<String, String> headers) {
        headers.forEach(this.headers::add);
    }

    @When("client calls {string} with body")
    public void clientCallsAddressWithBody(String address, String body) throws Exception {
        actions = mvc.perform(
            post(address)
                .headers(headers)
                .content(body)
        );
    }

    @Then("response status code is {int}")
    public void responseStatusCodeIs(Integer code) throws Exception {
        actions.andExpect(status().is(code));
    }

    @Then("response headers are")
    public void responseHeadersAre(Map<String, String> headers) {
        headers.forEach(this::assertHeader);
    }

    @Then("response body is")
    public void responseBodyIs(String body) throws Exception {
        actions.andExpect(content().json(body));
    }

    private void assertHeader(String name, String value) {
        try {
            actions.andExpect(header().string(name, value));
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }
}

File: application/src/test/kotlin/com/smtm/application/MvcStepdefs.kt
package com.smtm.application

import io.cucumber.java.ParameterType
import io.cucumber.java.en.Given
import io.cucumber.java.en.Then
import io.cucumber.java.en.When
import org.springframework.http.HttpHeaders
import org.springframework.http.HttpMethod
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.ResultActions
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders.request
import org.springframework.test.web.servlet.result.MockMvcResultMatchers.*

class MvcStepdefs(private val mvc: MockMvc) {

    private val headers: HttpHeaders = HttpHeaders()
    private lateinit var actions: ResultActions

    @Given("request headers are")
    fun requestHeadersAre(headers: Map<String, String>) {
        headers.forEach { (headerName, headerValue) -> this.headers.add(headerName, headerValue) }
    }

    @When("client performs a {httpMethod} {string} request with body")
    fun clientPerformsARequestWithBody(method: HttpMethod, endpoint: String, body: String) {
        actions = mvc.perform(
            request(method, endpoint)
                .headers(headers)
                .content(body)
        )
    }

    @When("client performs a {httpMethod} {string} request")
    fun clientPerformsARequest(method: HttpMethod, endpoint: String) {
        actions = mvc.perform(
            request(method, endpoint)
                .headers(headers)
        )
    }

    @Then("response status code is {int}")
    fun responseStatusCodeIs(code: Int) {
        actions.andExpect(status().`is`(code))
    }

    @Then("response headers are")
    fun responseHeadersAre(headers: Map<String, String>) {
        headers.forEach { (name, value) -> actions.andExpect(header().string(name, value)) }
    }

    @Then("response body is")
    fun responseBodyIs(body: String) {
        actions.andExpect(content().json(body))
    }

    @ParameterType("(GET|POST|PUT|DELETE|OPTIONS|TRACE|PATCH)")
    fun httpMethod(method: String): HttpMethod {
        return HttpMethod.valueOf(method)
    }

}
